language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codelenses/providers/DocsCodeLensesProvider.java

```diff
@@ -0,0 +1,171 @@
+        Class<? extends TopLevelNode> aTopLeveNodeClass = topLevelNode.getClass();
+        if (topLevelNode instanceof AnnotatableNode &&
+                (((AnnotatableNode) topLevelNode).getFlags() == null ||
+                        !((AnnotatableNode) topLevelNode).getFlags().contains(Flag.PUBLIC))) {
+            // If NOT Public; skip
+            return;
+        }
+        int line = topLevelNode.getPosition().getStartLine() - 1;
+        int col = topLevelNode.getPosition().getStartColumn();
+        boolean hasDocumentation = false;
+        for (Field field : aTopLeveNodeClass.getFields()) {
+            try {
+                if (field.getName().equals("markdownDocumentationAttachment")
+                        && field.get(topLevelNode) != null
+                        && field.get(topLevelNode) instanceof BLangMarkdownDocumentation) {
+                    BLangMarkdownDocumentation docs = (BLangMarkdownDocumentation) field.get(topLevelNode);
+                    line = docs.getPosition().getStartLine() - 1;
+                    col = docs.getPosition().getStartColumn();
+                    hasDocumentation = true;
+                }
+            } catch (IllegalAccessException e) {
+                // ignore
+            }
+        }
+        if (isDocumentable) {
+            Command command;
+            if (hasDocumentation) {
+                CommandArgument nameArg = new CommandArgument(CommandConstants.ARG_KEY_FUNCTION_NAME, nodeName);
+                List<Object> args = new ArrayList<>(Collections.singletonList(nameArg));
+                command = new Command("Preview Docs", "ballerina.showDocs", args);
+            } else {
+                CommandArgument nodeTypeArg = new CommandArgument(CommandConstants.ARG_KEY_NODE_TYPE, nodeType);
+                CommandArgument docUriArg = new CommandArgument(CommandConstants.ARG_KEY_DOC_URI, fileUri);
+                CommandArgument lineStart = new CommandArgument(CommandConstants.ARG_KEY_NODE_LINE,
+                                                                String.valueOf(line));
+                List<Object> args = new ArrayList<>(Arrays.asList(nodeTypeArg, docUriArg, lineStart));
+                command = new Command(CommandConstants.ADD_DOCUMENTATION_TITLE,
+                                      CommandConstants.CMD_ADD_DOCUMENTATION, args);
+            }
+            Position pos = new Position(line, col);
+            CodeLens lens = new CodeLens(new Range(pos, pos), command, null);
+            lenses.add(lens);
+        }
+    }
+
+    private Pair<String, String> resolveTopLevelType(TopLevelNode topLevelNode) {
+        if (topLevelNode instanceof BLangTypeDefinition) {
+            BLangTypeDefinition definition = (BLangTypeDefinition) topLevelNode;
+            if (definition.typeNode instanceof BLangObjectTypeNode) {
+                if (!((BLangObjectTypeNode) definition.typeNode).flagSet.contains(Flag.SERVICE)) {
```

Comment:
is it possible to merge these two if conditions to a single one?

Reply:
Since there's a casting involved; bringing it into a single line affects the readability. WDYT?